executor/simple.go

```diff
@@ -599,6 +603,35 @@ func (e *SimpleExec) executeBegin(ctx context.Context, s *ast.BeginStmt) error {
 	})
 }
 
+// ErrSavepointNotSupportedWithBinlog export for testing.
+var ErrSavepointNotSupportedWithBinlog = errors.New("SAVEPOINT is not supported when binlog is enabled")
+
+func (e *SimpleExec) executeSavepoint(s *ast.SavepointStmt) error {
+	sessVars := e.ctx.GetSessionVars()
+	txnCtx := sessVars.TxnCtx
+	if !sessVars.InTxn() && sessVars.IsAutocommit() {
+		return nil
+	}
+	if sessVars.BinlogClient != nil {
+		return ErrSavepointNotSupportedWithBinlog
```

Comment:
It has been explained in the design document. Since we not recoment use binlog after TiDB 4.0, and there are some new feature also not support binlog now, so here I also not support savepoint when binlog is enabled.